backend/services/roles.js

```diff
@@ -7,6 +8,52 @@ async function getAllRoles() {
   return Role.find();
 }
 
+/**
+ * Edit an existing role
+ */
+async function editRole(rawRole) {
+  const role = await Role.findById(rawRole._id);
+  if (role === null) {
+    throw ServiceError(404, "Role does not exist");
+  }
+  role.set(rawRole);
+  return role.save();
+}
+
+/**
+ * Create a new role
+ */
+async function createRole(rawRole) {
+  const role = await Role.findOne({ name: rawRole.name });
+  if (role !== null) {
+    throw ServiceError(409, `A role named '${rawRole.name}' already exists`);
+  }
+  return Role.create(rawRole);
+}
+
+/**
+ * Delete a role
+ */
+async function deleteRole(id) {
+  const role = await Role.findById(id);
+  if (role === null) {
+    throw ServiceError(404, "Role does not exist");
+  }
+
+  const usersWithRole = await User.find({ role }).exec();
+  await Promise.all(
+    usersWithRole.map((user) => async () => {
+      delete user.role;
```

Comment:
When a role gets deleted, their role is "Unassigned" which has the same permissions as Developer.

Reply:
Currently, the role attribute is optional, and newly created users have no role. Some of the existing codebase also assumes that a user's role can be null when checking for permissions etc. To make the Unassigned role behave like that, I would have to do the following:

* Make the role attribute required
* Remove null checks for the role attribute throughout the codebase
* Add a special case to make sure that the Unassigned role cannot be deleted
* Add code to create the Unassigned role if it does not already exist (e.g. when running for the first time), which might need to be separate from the mock data generation in order to work with the existing deployed database
* Change the role deletion code to use Unassigned instead of deleting the attribute

I agree that having an Unassigned role is a good idea in the long run, but I feel like it might work better in a separate PR. What do you think?

Reply:
Ah okay that makes sense, can we leave a TODO comment here then, just to be able to find it more easily in the future? I think he create an Unassigned role in our mock data, and I can talk to David to make sure that there is an Unassigned role in the production environment (I have no idea how the roles are created in production).

Reply:
Done!